test/create-sys-script.py

```diff
@@ -0,0 +1,202 @@
+
+
+def handle_file(outfile, path):
+    m = mode_rwx(os.lstat(path).st_mode)
+    f = open(path, "rb")
+    # read in as binary
+    b = f.read()
+    f.close()
+    # convert to string containing b'' representation
+    s = repr(b)
+    # strip leading b' and trailing '
+    s = s[2:-1]
+    # b'' contain newlines as \n, so convert back
+    s = s.replace(u'\\n', u'\n')
+    # we're placing it into a r'''''' so replace '
+    s = s.replace(u"'", u'\x27')
+    outfile.write(u"f(u'{}', u'{:o}', r'''{}''')\n".format(path, m, s))
+
+
+def process_sysdir(outfile):
+    for (dirpath, dirnames, filenames) in os.walk("sys"):
+        handle_dir(outfile, dirpath)
+        for d in dirnames:
+            path = os.path.join(dirpath, d)
+            if stat.S_ISLNK(os.lstat(path).st_mode):
+                handle_link(outfile, path)
+        for f in filenames:
+            path = os.path.join(dirpath, f)
+            mode = os.lstat(path).st_mode
+            if stat.S_ISLNK(mode):
+                handle_link(outfile, path)
+            elif stat.S_ISREG(mode):
+                handle_file(outfile, path)
+
+
+def verify_dir(tmpd, path_a):
+    path_b = os.path.join(tmpd, path_a)
+    mode_a = os.lstat(path_a).st_mode
+    mode_b = os.lstat(path_b).st_mode
+    if not stat.S_ISDIR(mode_b):
+        raise Exception(u"Not directory")
+    if mode_rwx(mode_a) != mode_rwx(mode_b):
+        raise Exception(u"Permissions mismatch")
+
+
+def verify_link(tmpd, path_a):
+    path_b = os.path.join(tmpd, path_a)
+    if not stat.S_ISLNK(os.lstat(path_b).st_mode):
+        raise Exception(u"Not symlink")
+    if os.readlink(path_a) != os.readlink(path_b):
+        raise Exception(u"Symlink dest mismatch")
+
+
+def verify_file(tmpd, path_a):
+    path_b = os.path.join(tmpd, path_a)
+    mode_a = os.lstat(path_a).st_mode
+    mode_b = os.lstat(path_b).st_mode
+    if not stat.S_ISREG(mode_b):
+        raise Exception(u"Not file")
+    if mode_rwx(mode_a) != mode_rwx(mode_b):
+        raise Exception(u"Permissions mismatch")
+    if not filecmp.cmp(path_a, path_b, shallow=False):
+        raise Exception(u"File contents mismatch")
+
+
+def verify_script(tmpd):
+    path = ""
+    try:
+        for (dirpath, dirnames, filenames) in os.walk("sys"):
+            path = dirpath
+            verify_dir(tmpd, path)
+            for d in dirnames:
+                path = os.path.join(dirpath, d)
+                if stat.S_ISLNK(os.lstat(path).st_mode):
+                    verify_link(tmpd, path)
+                for f in filenames:
+                    path = os.path.join(dirpath, f)
+                    mode = os.lstat(path).st_mode
+                    if stat.S_ISLNK(mode):
+                        verify_link(tmpd, path)
+                    elif stat.S_ISREG(mode):
+                        verify_file(tmpd, path)
+    except Exception as e:
+        print("FAIL on '{}'".format(path))
+        print(e)
+        exit(1)
```

Comment:
> Pfff, shouldn't we just verify the script once by checking that it recreates current directory properly?

yes, that's exactly what this does.  don't forget this is a meta-script, not the actual script.  it creates sys-script.py, and *that* is what is re-run every time we want to create the sys/ directory.  This verification is only done when sys-script.py is re-created, which is only when something is added/modified in the sys/ dir.

> Please no blanket except Exception

the tracebacks tell you where in this creation script it crashed, but that's not really useful because we aren't debugging the creation script (at this point of failure), we want to know which of the created files was created wrong.  Remember, these Exceptions are only run when we're creating the sys-script.py, which will only be done by developers who are adding/modifying sys/ file contents. But sure, we can pass on the exception info, I changed it to:

try:
...
except Exception as e:
    print("FAIL on '{}'".format(path))
    raise(e)